### PR TITLE
Add documentation for tests and features

### PR DESCRIPTION
**Pull request by shravan97**
Features in this PR : 
* Added comments for all tests and features in Pydoc format. You can now use `help(...)` to find out stuff about a method or a class

### CODE REVIEW

**Comment by shravan97** [comment]
@etejedor and other mentors, do you recommend making the comments under a class or a function definition as Numpy style ?

**Comment by etejedor** [comment]
@shravan97 what do you mean Numpy style?

File: PyRDF/CallableGenerator.py [hunk 1]
```diff
@@ -9,48 +9,6 @@ class CallableGenerator(object):
     def __init__(self, root_node):
         self.root_node = root_node
 
-    def _dfs(self, node, prev = 't', ops = []):
```
Comment (etejedor) [inline PyRDF/CallableGenerator.py]: Why is the removal of this function in this PR?

File: PyRDF/CallableGenerator.py [hunk 2]
```diff
@@ -62,22 +20,39 @@ def get_callable(self):
         self.root_node.graph_prune()
 
         def mapper(node_cpp, node_py=None):
+            """
+            The mapper callable that recurses
```
Comment (etejedor) [inline PyRDF/CallableGenerator.py]: I would not call it mapper in the documentation, actually at some point we should just name this function `callable`, since there will be no concept of mapper and reducer here, and the `get_reducer_callable` will be completely removed.

Comment (etejedor) [reply PyRDF/CallableGenerator.py]: Same comment applies for the docstring of `get_callable` and the one of `CallableGenerator`, please modify those as well.

File: PyRDF/CallableGenerator.py [hunk 3]
```diff
@@ -62,22 +20,39 @@ def get_callable(self):
...
+
+            """
             ## TODO : Somehow remove references to any Node object 
             ## for this to work on Spark
 
             return_vals = []
             return_nodes = []
 
             if not node_py:
+                # In the first recursive state
```
Comment (etejedor) [inline PyRDF/CallableGenerator.py]: Perhaps you can use two lines for this comment instead of 3?

File: PyRDF/CallableGenerator.py [hunk 4]
```diff
@@ -62,22 +20,39 @@ def get_callable(self):
...
+                # just set the current PyRDF
+                # node as the head node
                 node_py = self.root_node
             else:
+                # Execute the current operation using the output of the parent node (node_cpp)
                 node_cpp = getattr(node_cpp, node_py.operation.name)(*node_py.operation.args, **node_py.operation.kwargs)
+
                 if node_py.operation.op_type==Operation.Types.ACTION:
+                    # Collect all action nodes inorder to
```
Comment (etejedor) [inline PyRDF/CallableGenerator.py]: inorder -- in order

Comment (etejedor) [reply PyRDF/CallableGenerator.py]: and the whole comment in one line

File: PyRDF/CallableGenerator.py [hunk 5]
```diff
@@ -62,22 +20,39 @@ def get_callable(self):
...
+                    # return them
                     return_vals.append(node_cpp)
                     return_nodes.append(node_py)
 
             for n in node_py.next_nodes:
+                # Recurse through children and get
```
Comment (etejedor) [inline PyRDF/CallableGenerator.py]: One liner

File: PyRDF/Node.py [hunk 6]
```diff
@@ -5,13 +5,19 @@
...
+    operations graph. A Node houses an operation
+    and has references to child nodes. For details
+    on the types of operations supported, try :
+
+    import PyRDF
+    help(PyRDF.Operation)
 
     """
     def __init__(self, _get_head, operation):
-
```
Comment (etejedor) [inline PyRDF/Node.py]: Docstring of the constructor?

File: PyRDF/Node.py [hunk 7]
```diff
@@ -5,13 +5,19 @@
...
+
+    import PyRDF
+    help(PyRDF.Operation)
 
     """
     def __init__(self, _get_head, operation):
-        
+
         if _get_head is None:
+            # Function to get 'head' or 'root' Node
```
Comment (etejedor) [inline PyRDF/Node.py]: Let's just call it head node

File: PyRDF/Node.py [hunk 8]
```diff
@@ -22,27 +28,46 @@ def __init__(self, _get_head, operation):
         self.value = None
 
     def __getattr__(self, attr):
-        self._cur_attr = attr
+        # Interception of operation calls
```
Comment (etejedor) [inline PyRDF/Node.py]: Make it a proper docstring comment and elaborate a bit more: "Intercepts any call to an operation node and dispatches it by means of a call handler."

File: PyRDF/Node.py [hunk 9]
```diff
@@ -22,27 +28,46 @@ def __init__(self, _get_head, operation):
...
+
+        self._cur_attr = attr # Store new operation name
         return self._call_handler
 
     def _call_handler(self, *args, **kwargs):
```
Comment (etejedor) [inline PyRDF/Node.py]: Proper docstring here

File: PyRDF/Node.py [hunk 10]
```diff
@@ -22,27 +28,46 @@ def __init__(self, _get_head, operation):
...
             return Proxy(newNode)
 
         return newNode
     
     def graph_prune(self):
+        """
+        A Function to prune the nodes
+        of a PyRDF graph based on
+        user references
+
```
Comment (etejedor) [inline PyRDF/Node.py]: Elaborate a bit more: A node will be pruned if it has no children and the user application does not hold any reference to it.

File: PyRDF/Operation.py [hunk 11]
```diff
@@ -7,8 +7,35 @@ class Operation(object):
...
+    - Take
+    - Snapshot
+    - Foreach
+
     """
 
+    # An Enum Class member to categorize operations
     Types = Enum("Types", "ACTION TRANSFORMATION")
 
     def __init__(self, name, *args, **kwargs):
```
Comment (etejedor) [inline PyRDF/Operation.py]: docstring for the constructor

File: PyRDF/Operation.py [hunk 12]
```diff
@@ -22,6 +49,7 @@ def _classify_operation(self, name):
         Method to classify the given
         operation as action or 
         transformation
+
```
Comment (etejedor) [inline PyRDF/Operation.py]: Why this extra line?

File: PyRDF/Operation.py [hunk 13]
```diff
@@ -22,6 +49,7 @@ def _classify_operation(self, name):
...
         """
```
Comment (etejedor) [inline PyRDF/Operation.py]: It does not apply to this PR but: where are my requested `is_action` and `is_transformation` methods plus their corresponding use throughout the rest of the code to make it simpler? :smile: 
Can you make another PR with those changes?

File: PyRDF/Proxy.py [hunk 14]
```diff
@@ -10,20 +10,25 @@ class Proxy(object):
 
     """
```
Comment (etejedor) [inline PyRDF/Proxy.py:22]: I would reword the docstring of this class:
"""
Instances of Proxy act as futures of the result produced by some action.
They implement a lazy synchronization mechanism: when accessed for the first time, they trigger the execution of the whole RDataFrame graph.
"""

File: PyRDF/Proxy.py [hunk 15]
```diff
@@ -10,20 +10,25 @@ class Proxy(object):
...
 
+    # A Class member to store a backend
+    # object based on the configuration
+    # set by the user
     backend = None
 
     def __init__(self, action_node):
```
Comment (etejedor) [inline PyRDF/Proxy.py]: constructor docstring

File: PyRDF/Proxy.py [hunk 16]
```diff
@@ -10,20 +10,25 @@ class Proxy(object):
...
 
         self.action_node = action_node
 
     def __getattr__(self, attr):
+        # Intercept calls on the result
```
Comment (etejedor) [inline PyRDF/Proxy.py]: add a proper docstring

File: PyRDF/Proxy.py [hunk 17]
```diff
@@ -10,20 +10,25 @@ class Proxy(object):
...
+        # of an action node
 
         self._cur_attr = attr
         return self._call_handler
     
     def _call_handler(self, *args, **kwargs):
```
Comment (etejedor) [inline PyRDF/Proxy.py]: add docstring here

File: PyRDF/backend/Backend.py [hunk 18]
```diff
@@ -1,4 +1,10 @@
 class Backend(object):
+    """
+    A Generic Backend class (Interface)
+    that can be extended to define
+    specific 'execute()' functions
+
+    """
 
     def __init__(self, config={}):
```
Comment (etejedor) [inline PyRDF/backend/Backend.py]: Docstring for both init and execute here

Comment (etejedor) [reply PyRDF/backend/Backend.py]: Another comment, not for this PR, but for another one: there is the possibilty to specify abstract methods in Python, with the following syntax:
```
@abstractmethod
def execute(self, generator): pass
```
The advantage is that the check is not done at execution time like in the current solution. Can you verify that this works both in Python2 and Python3?

File: PyRDF/backend/Backend.py [hunk 19]
```diff
@@ -1,4 +1,10 @@
 class Backend(object):
+    """
+    A Generic Backend class (Interface)
```
Comment (etejedor) [inline PyRDF/backend/Backend.py]: """
Base class for RDataFrame backends.
Subclasses need to implement 'execute'.
"""

File: PyRDF/backend/Local.py [hunk 20]
```diff
@@ -2,19 +2,29 @@
...
         """
-        Execution of the event-loop
-        in local environment
+        Overridden abstract method from Backend
+        class to execute operations in local
+        environment
+
         """
-        mapper = generator.get_callable()
+        mapper = generator.get_callable() # Get a callable mapper
```
Comment (etejedor) [inline PyRDF/backend/Local.py]: Same comment as before, I would avoid referring to the callable as mapper.

Comment (etejedor) [reply PyRDF/backend/Local.py]: Just say callable.

File: PyRDF/backend/Local.py [hunk 21]
```diff
@@ -2,19 +2,29 @@
...
+    that executes the event-loop
+    (operations) locally
+
+    """
 
     def execute(self, generator):
         """
-        Execution of the event-loop
-        in local environment
+        Overridden abstract method from Backend
```
Comment (etejedor) [inline PyRDF/backend/Local.py]: """
Execute locally the current graph.
"""

File: PyRDF/backend/Local.py [hunk 22]
```diff
@@ -2,19 +2,29 @@
 from .Backend import Backend
 
 class Local(Backend):
+    """
```
Comment (etejedor) [inline PyRDF/backend/Local.py:5]: """
Backend that relies on the C++ implementation of RDataFrame to locally execute the current graph.
"""

File: tests/test_node.py [hunk 23]
```diff
@@ -1,36 +1,98 @@
 from PyRDF import *
 import unittest
```
Comment (etejedor) [inline tests/test_node.py:3]: Not all the tests in this file are documented, e.g. `test_dfs_graph_with_pruning_actions`, `test_dfs_graph_with_pruning_transformations`, ... Please make sure all of them are documented.

File: PyRDF/CallableGenerator.py (hunk 1 above)
Comment (shravan97) [reply PyRDF/CallableGenerator.py]: @etejedor , this function was a part of my previous implementation of Callable Generator. I thought I could utilize this commit for some clean up as well

File: PyRDF/Operation.py (hunk 13 above)
Comment (shravan97) [reply PyRDF/Operation.py]: @etejedor , oh yes. I'd created those in #14 , which I'd closed to split the commits. Forgot to make a separate PR during that time. Will one now surely :)

File: PyRDF/backend/Backend.py (hunk 18 above)
Comment (shravan97) [reply PyRDF/backend/Backend.py]: @etejedor sure. I'll verify and make a new PR for this 👍

**Comment by shravan97** [comment]
@etejedor , I'll make all the required changes 👍 

Btw, this is what I meant by Numpy style : [https://numpydoc.readthedocs.io/en/latest/format.html#class-docstring](https://numpydoc.readthedocs.io/en/latest/format.html#class-docstring) . Making it Numpy style would make it easier for us to move it to host it on [readthedocs.io](readthedocs.io) as well

File: PyRDF/CallableGenerator.py (hunk 1 above)
Comment (etejedor) [reply PyRDF/CallableGenerator.py]: Does the commit message say cleanup? :smile: Please separate in into another commit for clarity

Comment (shravan97) [reply PyRDF/CallableGenerator.py]: Yup @etejedor . Fair enough. I'll separate this into another commit 👍

**Comment by etejedor** [comment]
@shravan97 maybe you can use that format for class attributes? The rest of the documentation I would leave it as docstrings of classes and methods for now.
I do not know if we want to have our docs in readthedocs.io, but if we ever want to, is it hard to get it there from docstringed modules?

**Comment by shravan97** [comment]
@etejedor . I just checked out readthedocs.io again and it says that we need to create a separate markdown file for different modules. But then, in case of Numpy style docs, the docstrings would already be in Md format. But yes, I'll do what you suggested. Numpy style for classes and simple docstrings for methods 👍

**Comment by etejedor** [comment]
@shravan97 how are we doing with the comments proposed in this PR?

**Comment by shravan97** [comment]
@etejedor , almost done. I'm just taking some time to make the docs a little more extensive. This should ve done in an hour or two :+1:

**Comment by shravan97** [comment]
@etejedor , this is done now 👍 . I made the docs more consistent for all non test files and wrote them a little extensively

**Comment by shravan97** [comment]
And @etejedor , for remove the `_dfs` function, I'll make a new PR, once this is merged 👍

File: PyRDF/CallableGenerator.py [hunk 24]
```diff
@@ -2,18 +2,32 @@
...
     """
-    Class that generates
-    mapper and reducer functions
+    Class that generates a callable
+    to parse a PyRDF graph.
+
+    Attributes
+    ----------
+    root_node
+        Head node a PyRDF graph.
```
Comment (etejedor) [inline PyRDF/CallableGenerator.py]: of a PyRDF graph?

File: PyRDF/CallableGenerator.py [hunk 25]
```diff
@@ -2,18 +2,32 @@
...
 
     """
     def __init__(self, root_node):
+        """
+        Creates a new `CallableGenerator`.
+
+        Parameters
+        ----------
+        root_node
+            Head node a PyRDF graph.
```
Comment (etejedor) [inline PyRDF/CallableGenerator.py]: of a PyRDF graph?

File: PyRDF/CallableGenerator.py [hunk 26]
```diff
@@ -2,18 +2,32 @@
...
 class CallableGenerator(object):
     """
-    Class that generates
-    mapper and reducer functions
+    Class that generates a callable
+    to parse a PyRDF graph.
+
+    Attributes
+    ----------
+    root_node
```
Comment (etejedor) [inline PyRDF/CallableGenerator.py:10]: We might as well call the member `head_node`

File: PyRDF/CallableGenerator.py [hunk 27]
```diff
@@ -2,18 +2,32 @@
...
+        Head node a PyRDF graph.
 
     """
     def __init__(self, root_node):
```
Comment (etejedor) [inline PyRDF/CallableGenerator.py]: and this parameter, also `head_node`, to be consistent with the docs.

File: PyRDF/CallableGenerator.py [hunk 28]
```diff
@@ -53,31 +67,65 @@ def _dfs(self, node, prev = 't', ops = []):
...
-        Function that converts a given
-        graph onto a mapper function and 
-        returns the same
+        Converts a given graph into
+        a callable and returns the same.
+
+        Returns
+        -------
+        function
+            The callable function that traverses
```
Comment (etejedor) [inline PyRDF/CallableGenerator.py]: callable function is perhaps redundant, I would just leave either function or callable.

File: PyRDF/Node.py [hunk 29]
```diff
@@ -5,44 +5,140 @@
...
 
     """
     def __init__(self, _get_head, operation):
-        
+        """
+        Creates a new `Node` based on the 'operation'.
+
+        Parameters
+        ----------
+        _get_head : function
```
Comment (etejedor) [inline PyRDF/Node.py:40]: Perhaps this parameter should not have the leading `_`, since it has to appear (as parameter) in the documentation.

File: PyRDF/Node.py [hunk 30]
```diff
@@ -5,44 +5,140 @@
...
+            A method to handle an operation call to the
+            current node.
+
+        """
+
+        self._cur_attr = attr # Stores new operation name
         return self._call_handler
 
     def _call_handler(self, *args, **kwargs):
+        """
```
Comment (etejedor) [inline PyRDF/Node.py]: This is an internal method, so it should not have docstring attached. The documentation should be in the form of comments.

File: PyRDF/Node.py [hunk 31]
```diff
@@ -5,44 +5,140 @@
...
+        Handles an operation call to the current node.
+
+        Parameters
+        ----------
+        *args
+            Variable length argument list for the incoming
+            operation.
+
+        **kwargs
+            Arbitrary keyword arguments for the incoming
```
Comment (etejedor) [inline PyRDF/Node.py]: Why arbitrary?

File: PyRDF/Operation.py [hunk 32]
```diff
@@ -3,25 +3,92 @@
...
+        ----------
+        name : str
+            Name of the current operation
+
+        *args
+            Variable length argument list for the current
+            operation.
+
+        **kwargs
+            Arbitrary keyword arguments for the current
```
Comment (etejedor) [inline PyRDF/Operation.py]: Arbitrary?

File: PyRDF/Proxy.py [hunk 33]
```diff
@@ -3,27 +3,77 @@
...
+        Handles an operation call to the current node.
+
+        Parameters
+        ----------
+        *args
+            Variable length argument list for the incoming
+            operation.
+
+        **kwargs
+            Arbitrary keyword arguments for the incoming
```
Comment (etejedor) [inline PyRDF/Proxy.py]: Arbitrary

File: PyRDF/RDataFrame.py [hunk 34]
```diff
@@ -5,28 +5,52 @@
...
     
     """
-    The Python equivalent
-    of ROOT C++'s RDataFrame 
-    data structure
-
-    Supported constructor arguments :
-    - Number of entries (Int)
+    The Python equivalent of ROOT C++'s
+    RDataFrame data structure.
```
Comment (etejedor) [inline PyRDF/RDataFrame.py]: RDataFrame class instead of data structure

File: PyRDF/RDataFrame.py [hunk 35]
```diff
@@ -5,28 +5,52 @@
...
+
+    Attributes
+    ----------
+    args
+        A list of arguments that were provided to
+        construct the RDataFrame object.
+
+    Supported constructor arguments
+    -------------------------------
+    - number of entries (Int)
```
Comment (etejedor) [inline PyRDF/RDataFrame.py]: lowercase `int`

File: PyRDF/RDataFrame.py [hunk 36]
```diff
@@ -43,9 +67,8 @@ def __init__(self, *args):
 
     def _get_vector_from_list(self, arg):
         """
-        Helper method to convert
-        a python list of strings
-        to a vector
+        converts a python list of strings
```
Comment (etejedor) [inline PyRDF/RDataFrame.py]: Internal method, docstring should be in the form of comments.
Also uppercase `Converts`.

File: PyRDF/CallableGenerator.py (hunk 24 above)
Comment (shravan97) [reply PyRDF/CallableGenerator.py]: @etejedor , I meant Python RDataFrame graph. Since this isn't the CPP RDF graph, I made it Python explicitly

File: PyRDF/CallableGenerator.py (hunk 26 above)
Comment (shravan97) [reply PyRDF/CallableGenerator.py:10]: @etejedor , can I include this along with the PR to remove `_dfs` ?

That PR will be completely for cleaning up and refactoring variable names. So I guess it'd be better to include this in that PR. Is that okay ?

File: PyRDF/Node.py (hunk 29 above)
Comment (shravan97) [reply PyRDF/Node.py:40]: And this as well. Can I include this as a part of the clean up PR ?

File: PyRDF/CallableGenerator.py (hunk 24 above)
Comment (shravan97) [reply PyRDF/CallableGenerator.py]: Shall then change this to " Python RDataFrame graph" then @etejedor ?

File: PyRDF/Node.py (hunk 31 above)
Comment (shravan97) [reply PyRDF/Node.py]: I've written arbitrary because the keyword arguments could be anything. But yes, I think it looks much better without "Arbitrary". I'll just leave it as "Keyword arguments for ..... "

File: PyRDF/CallableGenerator.py (hunk 24 above)
Comment (JavierCVilla) [reply PyRDF/CallableGenerator.py]: I think he means you missed **of**  right before **a PyRDF**

Comment (shravan97) [reply PyRDF/CallableGenerator.py]: Oh yes. I'm sorry that I didn't notice it. Thank @JavierCVilla 😬

File: PyRDF/CallableGenerator.py (hunk 26 above)
Comment (etejedor) [reply PyRDF/CallableGenerator.py:10]: Sure

File: PyRDF/Node.py (hunk 29 above)
Comment (etejedor) [reply PyRDF/Node.py:40]: Yes sure

File: PyRDF/Node.py (hunk 30 above)
Comment (shravan97) [reply PyRDF/Node.py]: But @etejedor , what if the user tries something like this : 

```python
r = PyRDF.RDataFrame(10)
help(r.Define)
```

Comment (shravan97) [reply PyRDF/Node.py]: In that case @etejedor , shall I remove the trailing `_` and change it to `call_handler` ?

Comment (etejedor) [reply PyRDF/Node.py]: That is another topic. Ideally `help(r.Define)` should show us the C++ documentation for the real `Define`. That is currently not the case and we should hook in that redirection, but I would leave that aside for now.
I would then just keep `_call_handler` with the underscore, because it is internal. Therefore, its documentation needs to be a comment.

Comment (shravan97) [reply PyRDF/Node.py]: Ok sure @etejedor 👍

**Comment by shravan97** [comment]
@etejedor , could you review now ?
